### PR TITLE
Update to Metadata v5

### PR DESCRIPTION
**Pull request by Stefie**
Addresses #988

### CODE REVIEW

File: packages/api-derive/package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@polkadot/api-derive",
-  "version": "0.81.0-beta.7",
+  "version": "0.81.0",
```
Comment (jacogr) [inline packages/api-derive/package.json]: Why this change? Doing so will make the merge publish fail.

Comment (jacogr) [reply packages/api-derive/package.json]: i.e. it is now going to try and re-publich 0.81.0-beta.1 (over the previous already-published version), and fail.

Comment (Stefie) [reply packages/api-derive/package.json]: Yeah, right. It should be `0.82.0`. I just looked at @amaurymartiny PR from 2 months ago when he updated Metadata from v4 to v5 and I just copied his versioning mishap https://github.com/polkadot-js/api/pull/857/files :|

Comment (jacogr) [reply packages/api-derive/package.json]: Ok, the issue however is - we are not doing a stable release here. We have not tested this and any of the other changes sufficiently - so no bumps, we are not in a position to do this. Once we do, it goes out everywhere.,

File: lerna.json [hunk 2]
```diff
@@ -9,5 +9,5 @@
   "packages": [
     "packages/*"
   ],
-  "version": "0.81.0-beta.8"
+  "version": "0.82.0"
```
Comment (jacogr) [inline lerna.json]: As per the comments, this is a PR, not a release merge for release. Please revert the version bumps completely.

Comment (jacogr) [reply lerna.json]: To expand -

- we don't release until there has been sufficient testing of the beta package in apps (and there are some stuff that are still up-in-the air and may have impact of functionality, i.e. the contracts structure changes)

- we don't make api releases where the API depends on a non-stable common package tree (as it does here) - only stable versions of common for API stable releases

Comment (Stefie) [reply lerna.json]: Ok, I didn't know that. Happy to remove the version bumps, then. 
What exactly do you mean by:
>  non-stable common package tree

The fact that I still have the `polkadot-js/common` modules included as beta? So for a release, I would need to bump up all `@polkadot` modules to their next major version?

Comment (Stefie) [reply lerna.json]: The API is broken with the substrate master atm, though. So it could be good to get out a smaller release before getting all the stuff in that is still in the air.

Comment (jacogr) [reply lerna.json]: All the "stuff in the air" is merged already.

File: packages/types/src/Metadata/v5/Storage.ts [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+import Enum from '../../codec/Enum';
+import Struct from '../../codec/Struct';
+import Vector from '../../codec/Vector';
+import Bool from '../../primitive/Bool';
+import Bytes from '../../primitive/Bytes';
+import StorageHasher from '../../primitive/StorageHasher';
+import Text from '../../primitive/Text';
+import Type from '../../primitive/Type';
+import { PlainType, StorageFunctionModifier } from '../v3/Storage';
```
Comment (amaurymartiny) [inline packages/types/src/Metadata/v5/Storage.ts]: Better import/export from v4, so that we know what changed in each version in one glance

File: packages/types/src/Metadata/v5/Storage.ts [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+
+// Re-export classes that haven't changed between V3 and V5
+export {
+  PlainType,
+  StorageFunctionModifier
+};
+
+export class MapType extends Struct {
```
Comment (amaurymartiny) [inline packages/types/src/Metadata/v5/Storage.ts]: Can export directly the v4 one, nothing changed in this

File: packages/types/src/Metadata/v4/toV5.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        return new StorageHasher('Twox64Concat');
+      default:
+        throw new Error(`Invalid Storage hasher: ${text.toString()}`);
+    }
+  }
+
+  const [newType, index] = type.isPlainType
+    ? [type, 0]
+    : type.isMap
+      ? [new MapType({
```
Comment (amaurymartiny) [inline packages/types/src/Metadata/v4/toV5.ts]: I think we can replace this by `[type, 1]` (if TS doesn't complain)

File: packages/types/src/Metadata/v4/toV5.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        hasher: new StorageHasher('Twox128'),
+        key: type.asMap.key,
+        value: type.asMap.value,
+        isLinked: type.asMap.isLinked
+      }), 1]
+      : [new DoubleMapType({
+        hasher: new StorageHasher('Twox128'),
```
Comment (amaurymartiny) [inline packages/types/src/Metadata/v4/toV5.ts]: here it needs to be type.asDoubleMap.hasher

File: packages/types/src/Metadata/v5/Calls.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+// Copyright 2017-2019 @polkadot/types authors & contributors
+// This software may be modified and distributed under the terms
+// of the Apache-2.0 license. See the LICENSE file for details.
+
+// Re-export classes that haven't changed between V3 and V4
```
Comment (amaurymartiny) [inline packages/types/src/Metadata/v5/Calls.ts]: v4 v5

File: packages/types/src/Metadata/v5/Events.ts [hunk 8]
```diff
@@ -0,0 +1,6 @@
+// Copyright 2017-2019 @polkadot/types authors & contributors
+// This software may be modified and distributed under the terms
+// of the Apache-2.0 license. See the LICENSE file for details.
+
+// Re-export classes that haven't changed between V3 and V4
```
Comment (amaurymartiny) [inline packages/types/src/Metadata/v5/Events.ts]: v4 v5

File: packages/types/src/Metadata/v4/toV5.ts (hunk 6 above)
Comment (Stefie) [reply packages/types/src/Metadata/v4/toV5.ts]: Thanks, changed.

File: packages/types/src/Metadata/v4/toV5.ts (hunk 5 above)
Comment (Stefie) [reply packages/types/src/Metadata/v4/toV5.ts]: `type.asMap` did the trick

File: packages/types/src/Metadata/v5/Storage.ts (hunk 3 above)
Comment (Stefie) [reply packages/types/src/Metadata/v5/Storage.ts]: done.

File: packages/type-storage/src/fromMetadata/getHasher.ts [hunk 9]
```diff
@@ -2,41 +2,19 @@
 // This software may be modified and distributed under the terms
 // of the Apache-2.0 license. See the LICENSE file for details.
 
-import { Text } from '@polkadot/types';
 import { StorageHasher } from '@polkadot/types/primitive';
 import { blake2AsU8a, xxhashAsU8a } from '@polkadot/util-crypto';
 
-type HasherInput = string | Buffer | Uint8Array;
```
Comment (Stefie) [inline packages/type-storage/src/fromMetadata/getHasher.ts]: Do we need the `string` in the type definition anywhere else than here? https://github.com/polkadot-js/api/blob/master/packages/type-storage/src/fromMetadata/createFunction.ts#L82

**Comment by Stefie** [comment]
I ran the e2e tests in `api`, `api-derive`, `rpc-core` and `rpc-rx`. Anything else I should test that I'm not aware of?

**Comment by jacogr** [comment]
For these we would just need to trust the tests :) So just make sure you have tested against both new (substrate master) and Alex - i.e. old and new, and it should be all-ok.

**Comment by Stefie** [comment]
I did and I just dumped the compiled files in the Polkadot Apps `node_modules`: Everything is working again and I'm getting no errors with Alex, Embric Elm and  the latest substrate.

So from my side we're good to go...

File: packages/types/src/Metadata/MetadataVersioned.ts [hunk 10]
```diff
@@ -242,4 +262,21 @@ export default class MetadataVersioned extends Struct {
 
     return this._convertedV4;
   }
+
+  /**
+   * @description Returns the wrapped values as a V5 object
+   */
+  get asV5 (): MetadataV5 {
+    assert(this.metadata.version <= 5, `Cannot convert metadata from v${this.metadata.version} to v4`);
```
Comment (amaurymartiny) [inline packages/types/src/Metadata/MetadataVersioned.ts]:
```suggestion
    assert(this.metadata.version <= 5, `Cannot convert metadata from v${this.metadata.version} to v5`);
```

Comment (amaurymartiny) [reply packages/types/src/Metadata/MetadataVersioned.ts]: Also just above, v3->v4

**Comment by Stefie** [comment]
@jacogr what's your requested change?

**Comment by jacogr** [comment]
Still missing this one (unrelated, not touched), but as @amaurymartiny suggested, needs a fix - "Also just above, v3->v4"

**Comment by Stefie** [comment]
Um, sorry, but I don't understand your comment? 
What do you mean by "Still missing this one (unrelated, not touched)"? 
And where do you have the "v3->v4" ? 

@amaurymartiny 's "v4 -> v5" PR/ suggestion is already merged.

**Comment by jacogr** [comment]
https://github.com/polkadot-js/api/pull/991#discussion_r293343422
